File: backend/app/engine/clustering.py
"""Мульти-ОКС: кластеризация под общий ствол (§2.11 ТЗ, Спринт 2).

Перспективные ОКС, чьи точки подключения ближе cluster_radius_m,
рассматриваются как кандидаты на СОВМЕСТНОЕ подключение: один ствол
от точки врезки до камеры разветвления, от неё — ветви к каждому ОКС.
Эвристика по близости (Steiner-задача в общем виде NP-сложная);
раздельное подключение — допустимый результат и случается автоматически,
когда кластер состоит из одного ОКС.
"""

from __future__ import annotations

from .model import Building


def cluster_buildings(buildings: dict[str, Building], radius_m: float) -> list[list[Building]]:
    """Жадная кластеризация по близости точек подключения (union-find)."""
    items = list(buildings.values())
    parent = list(range(len(items)))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a: int, b: int) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    anchors = [b.anchor for b in items]
    for i in range(len(items)):
        for j in range(i + 1, len(items)):
            if anchors[i].distance(anchors[j]) <= radius_m:
                union(i, j)

    clusters: dict[int, list[Building]] = {}
    for idx, b in enumerate(items):
        clusters.setdefault(find(idx), []).append(b)

    # Крупные кластеры в начало — им важнее хорошая точка врезки
    return sorted(clusters.values(), key=lambda c: -len(c))

File: app/src/main/java/ru/heatnet/dit/engine/core/Clustering.java
package ru.heatnet.dit.engine.core;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Кластеризация точек подключения под общий ствол (§2.1: несколько точек
 * могут использовать общий участок сети).
 *
 * Точки oks_connection_point, расположенные ближе cluster_radius_m друг
 * от друга, — кандидаты на СОВМЕСТНОЕ подключение: один ствол от места
 * присоединения до камеры разветвления, от неё — ветви к каждой точке.
 */
public final class Clustering {

    private Clustering() {
    }

    /** Жадная кластеризация по близости точек подключения (union-find). */
    public static List<List<Model.ConnectionTarget>> clusterTargets(
            Map<String, Model.ConnectionTarget> targets, double radiusM) {
        List<Model.ConnectionTarget> items = new ArrayList<>(targets.values());
        int n = items.size();
        int[] parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (items.get(i).anchor().distance(items.get(j).anchor()) <= radiusM) {
                    union(parent, i, j);
                }
            }
        }
        Map<Integer, List<Model.ConnectionTarget>> clusters = new LinkedHashMap<>();
        for (int i = 0; i < n; i++) {
            clusters.computeIfAbsent(find(parent, i), k -> new ArrayList<>()).add(items.get(i));
        }
        // Крупные кластеры в начало — им важнее хорошая точка присоединения
        List<List<Model.ConnectionTarget>> out = new ArrayList<>(clusters.values());
        out.sort((a, b) -> Integer.compare(b.size(), a.size()));
        return out;
    }

    private static int find(int[] parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private static void union(int[] parent, int a, int b) {
        int ra = find(parent, a), rb = find(parent, b);
        if (ra != rb) {
            parent[ra] = rb;
        }
    }
}
